File: tests/typesafety/python/exercise_lookup_misuse.py
# expect: fails
# title: A Result where greet needs the plain user id, rejected by the checker
# misuse-of: exercise_lookup_solution
from dataclasses import dataclass


@dataclass(frozen=True)
class Ok[T]:
    value: T


@dataclass(frozen=True)
class Err:
    reason: str


type Result[T] = Ok[T] | Err


def find_user_id(login: str) -> Result[str]:
    return Ok("u7") if login == "ada" else Err(f"no user for login {login!r}")


def greet(user_id: str) -> str:
    return f"welcome {user_id}"


banner = greet(find_user_id("ada"))

File: tests/typesafety/typescript/exercise_lookup_misuse.ts
// expect: fails

type Ok<T> = { readonly kind: "ok"; readonly value: T };
type Err = { readonly kind: "err"; readonly reason: string };
type Result<T> = Ok<T> | Err;

function findUserId(login: string): Result<string> {
  return login === "ada" ? { kind: "ok", value: "u7" } : { kind: "err", reason: `no user for login ${login}` };
}

function greet(userId: string): string {
  return `welcome ${userId}`;
}

export const banner = greet(findUserId("ada"));
